Add sortable param to allow named filters agg return buckets as an array of objects

amooncake opened the pull request:
Named filters with bucket sort currently return object, which may not be
reliably supported by JSON since it doesn't guarantee the order of
elements in the object. In this case, add a sortable parameter to the
filters agg, which will be ignored by the anonymous filters, the buckets
will be returned in an array of objects when it is true.

Closes #83957

amooncake committed:
Change return type when named filters aggs with bucket sort

Named filters with bucket sort currently return object, which
may not be reliably supported by JSON since it doesn't guarantee
the order of elements in the object. In this case, return them as
an array of object.

amooncake committed:
add changelog

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketSortPipelineAggregator.java:
```diff
@@ -78,7 +79,13 @@ public InternalAggregation reduce(InternalAggregation aggregation, AggregationRe
+            return new InternalFilters(internalFilters.getName(), internalFilters.getBuckets(), false, internalFilters.getMetadata());
+        }
+
+        return internalAggregation;
```
I'm not super comfortable having one aggregation cause another aggregation to change it's rendering like this. I get that it's a much more sensible way to do it, but it feels weird. I'd prefer to have this be something you configure on the agg up front. Over [here](https://github.com/elastic/elasticsearch/issues/83957#issuecomment-1083317711) we talked about letting folks configure the `keyed` parameter - but I'm not really sure that makes sense....

Both this solution and overriding the `keyed` parameter kind of have the opposite problem - if you don't name the filters and then you sort then you can't figure out which one is which.

It all sort of makes me wonder - why point the bucket_sort pipeline at the filters agg at all? It sort of breaks it.

Finally, I'm a bit worried about changing this json output without some kind of extra parameter because it's *technically* breaking. I mean, the way we're outputting these now is kind of garbage in combination with the `bucket_sort` pipeline agg. But so many folks use Elasticsearch right now, any change in json format is likely to break someone.

So I wonder if the "right" way to fix this is to add a new parameter to the `filters` agg that causes it to output the results in this way. Maybe a `ordered_buckets: true` or  something. I don't know. I'm not good at names. And, yeah, it's lame to need a new parameter. But it's paranoid. And in this case I think that paranoia is good.

amooncake replied:
I agreed, and then I added a keyed_bucket_in_array param as you suggested to avoid judging this in bucket sort.

Now the user can directly modify the return type regardless of whether there is a bucket sort.

amooncake committed:
Add keyed_bucket_in_array param in filters aggs to change return type

Named filters with bucket sort currently return object, which
may not be reliably supported by JSON since it doesn't guarantee
the order of elements in the object. In this case, set the param
to true to return an array of object.

nik9000 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/220_filters_bucket.yml:
```diff
@@ -137,6 +137,40 @@ setup:
+          aggs:
+            the_filter:
+              filters:
+                keyed_bucket_in_array: true
```
I don't particularly like the name of this parameter. But I don't really know of a better name. @not-napoleon, do you have an opinion here?

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregationBuilder.java:
```diff
@@ -123,6 +126,7 @@ public FiltersAggregationBuilder(StreamInput in) throws IOException {
         }
         otherBucket = in.readBoolean();
         otherBucketKey = in.readString();
+        keyedBucketInArray = in.readBoolean();
```
This'll probably cause something like:
```
./gradlew -p qa/mixed-cluster/ -Dtests.method='*filter*'
```

to fail. The fix it to change it to something like:
```
keyedBucketInArray = in.getVersion().before(Version.V_8_5_0) ? false : in.readBoolean();
```

And to skip the write on the other side if the version is too old.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/InternalFilters.java:
```diff
@@ -148,6 +157,7 @@ public InternalFilters(StreamInput in) throws IOException {
     @Override
     protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeBoolean(keyed);
+        out.writeBoolean(keyedBucketInArray);
```
This needs the same guard as I mentioned above, sadly.

nik9000 commented:
@elasticsearchmachine, ok to test.

amooncake committed:
Add version check for the new param

amooncake commented:
@elasticsearchmachine update branch

nik9000 commented:
@elasticsearchmachine update branch

nik9000 commented:
> @elasticsearchmachine update branch

I think it needs employees.

nik9000 commented:
@elasticsearchmachine, update branch please

amooncake commented:
What a coincidence, I'll update manually.

nik9000 commented:
> What a coincidence, I'll update manually.

:+1:

amooncake committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into named_filters_aggs

nik9000 commented:
Could you rename the new parameter to `sortable`?

nik9000 commented:
It's not a great name. But maybe a little better than other one...

amooncake committed:
Rename the new parameter to sortable

amooncake committed:
Fix format

amooncake committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into named_filters_aggs

nik9000 reviewed (approved):
Looks good to me. Let's get this merged. Are you interested in putting together some docs for this? That's it's own process. If not I can put something together in a bit.

amooncake committed:
Update filters-aggregation doc

amooncake committed:
Merge branch 'elastic:main' into named_filters_aggs

amooncake committed:
Fix doc linking

amooncake committed:
Merge branch 'elastic:main' into named_filters_aggs

amooncake commented:
Preview link: https://elasticsearch_89256.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-filters-aggregation.html#sortable-buckets

nik9000 commented:
> Preview link: https://elasticsearch_89256.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-filters-aggregation.html#sortable-buckets

Nice!

I'm trying to get some folks from our clients team to have a look at this to make sure it doesn't break them.

nik9000 commented:
Hey! I talked this over with the clients team and they felt like it'd be quite a lot of overhead for them to support the different "shape" here. Even though it's better, it may not be worth doing just because of their overhead. Especially when folks can sort on the client side. I'm going to invoke @qhoxie and @not-napoleon to have another look here and try and figure out if this is worth the overhead on their side. Sorry for the delay here. And for the uncertainty.

sethmlarson commented:
@amooncake Apologies for the delay, I've provided a summary of my thoughts on how to best accomplish this change here: https://github.com/elastic/elasticsearch/issues/83957#issuecomment-1304115505

If you're still interested in fixing this issue I believe the changes described above aren't far off what you have now, only need a little tweaking.

amooncake committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into named_filters_aggs

amooncake committed:
convert `sortable` to `keyed`, add `key` in bucket

amooncake committed:
Fix test case

amooncake committed:
Update min version for `keyed` param

amooncake commented:
Since the `keyed` parameter is already used in the source code to indicate the keyed filters, the newly added `keyed` parameter in the DSL will be parsed as `keyedBucket` in the source code.
https://github.com/elastic/elasticsearch/blob/91a2d5ddf881575e0e33cb1c3d2a81cd4942f610/server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregationBuilder.java#L42-L51
https://github.com/elastic/elasticsearch/blob/91a2d5ddf881575e0e33cb1c3d2a81cd4942f610/server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregationBuilder.java#L290-L291
The current implementation feels a bit weird and confusing. But I don't know of a better name or implementation. @nik9000 @sethmlarson Could you help me with this?

nik9000 reviewed (approved):
OK! Sorry for the huge delay here. It looks right except for the versions are stale. I'll update it.

nik9000 commented:
It is weird and confusing, yeah. But there are worse things, I guess.

nik9000 committed:
Merge branch 'main' into pr_89256

nik9000 committed:
New version!

nik9000 commented:
@not-napoleon could you review this as well? I feel like it's blurred a bit for me and would appreciate another set of eyes on it.

not-napoleon reviewed (approved):
LGTM

nik9000 reviewed (commented):
There seems to be a backwards compatibility issue. Checking.

nik9000 committed:
better?

nik9000 commented:
Thanks for all of your patience @amooncake.

amooncake commented:
Thanks @nik9000 !
